### PR TITLE
ledger: fix accountsWriting WaitGroup use

### PR DESCRIPTION
**Pull request by algorandskiy**
## Summary

This is a followup change for #5876

* accountsWriting used to provide an auxiliary waiting primitive
   for account flushing after adding blocks. The main use case is in tests
   to get accounts flushed and queryable after injection blocks.
* Unlike a correct way of using WaitGroups, Add and Wait were calling
   in different goroutines causing rare data races.
* This fix replaces the WaitGroup with a atomic counter with the same
   semantics as a WaitGroup but with spinning instead of Wait call.

## Test Plan

Existing unit tests

### CODE REVIEW

File: ledger/acctonline_test.go [hunk 1]
```diff
@@ -38,17 +38,10 @@ import (
...
-	dcc := &deferredCommitContext{
-		deferredCommitRange: deferredCommitRange{
-			lookback: maxLookback,
-		},
-	}
-	cdr := ml.trackers.produceCommittingTask(rnd, ml.trackers.dbRound, &dcc.deferredCommitRange)
-	if cdr != nil {
+	dcc := ml.trackers.produceCommittingTask(rnd, ml.trackers.dbRound, maxLookback)
+	if dcc != nil {
 		func() {
```
Comment (jannotti) [inline ledger/acctonline_test.go:43]: What is going on with this anonymous function being created and immediately invoked?  I don't understand what that is accomplishing here.

**Review by jannotti — COMMENTED** [review]
I don't understand the basic problem, or at least I don't understand how changing to atomic would change anything.  I'm speculating that the race being considered is what happens when you try to use a go routine that performs the `Add()` in the goroutine being started. So if you `Wait`, and the goroutine hasn't been scheduled yet, it seems like you're done.  But if that's the problem, them an atomic int wold have the same problem.  So do I misunderstand the problem, or is there something else that makes it ok to swap in an atomic int?

File: ledger/acctonline_test.go (hunk 1 above)
Comment (algorandskiy) [reply ledger/acctonline_test.go:43]: Agree, this function does not make sense since these is no defer (like RAII patttern)

**Comment by algorandskiy** [comment]
> I don't understand the basic problem, or at least I don't understand how changing to atomic would change anything.

Look at this data race:
```
==================
WARNING: DATA RACE
Read at 0x00c0021f8a78 by goroutine 3257:
  runtime.raceread()
      <autogenerated>:1 +0x24
  github.com/algorand/go-algorand/ledger.(*trackerRegistry).scheduleCommit()
      /opt/cibuild/project/ledger/tracker.go:456 +0x364
  github.com/algorand/go-algorand/ledger.(*trackerRegistry).committedUpTo()
      /opt/cibuild/project/ledger/tracker.go:395 +0x1d8
  github.com/algorand/go-algorand/ledger.(*Ledger).notifyCommit()
      /opt/cibuild/project/ledger/ledger.go:455 +0x19c
  github.com/algorand/go-algorand/ledger.(*blockQueue).syncer()
      /opt/cibuild/project/ledger/blockqueue.go:166 +0x813
  github.com/algorand/go-algorand/ledger.(*blockQueue).start.func3()
      /opt/cibuild/project/ledger/blockqueue.go:92 +0x39

Previous write at 0x00c0021f8a78 by goroutine 3240:
  runtime.racewrite()
      <autogenerated>:1 +0x24
  github.com/algorand/go-algorand/ledger.(*trackerRegistry).waitAccountsWriting()
      /opt/cibuild/project/ledger/tracker.go:476 +0x2f7
  github.com/algorand/go-algorand/ledger.triggerTrackerFlush()
      /opt/cibuild/project/ledger/ledger_test.go:1469 +0x29b
  github.com/algorand/go-algorand/ledger.TestLedgerCatchpointSPVerificationTracker()
      /opt/cibuild/project/ledger/ledger_test.go:3170 +0x6ac
  testing.tRunner()
      /opt/cibuild/.gimme/versions/go1.20.7.linux.amd64/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /opt/cibuild/.gimme/versions/go1.20.7.linux.amd64/src/testing/testing.go:1629 +0x47
```

One thread was calling Add and another Wait. Apparently, WaitGroup does not handle this right.

File: ledger/acctonline_test.go (hunk 1 above)
Comment (algorandskiy) [reply ledger/acctonline_test.go]: fixed/removed

**Comment by algorandskiy** [comment]
Discussed offline: the waitgroup is used across 3 gorouting (block syncer calls Add, commit syncer calls Done, and the main calls Wait). Decided even with atomic counters there is still space for race condition (but not data race), and probably need other approach).

**Pull request closed without merging**